design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,433 @@
+specifics of this effort will be left to be determined as part of the
+implementation. In any case, when deleting used or expired SUTs, we must not
+cascade the delete since this would delete any related audit logs.
+
+##### Code Verifier and Code Challenge
+
+As an additional layer of security, the client will generate a code verifier,
+which can be a Guid or other cryptographically secure random string. The code
+will then be hashed and sent along with the login request to be associated
+with the generated SUT. This hashed code verifier, known as the code
+challenge, will be stored by Conjur along with the SUT. When the client
+attempts to authenticate with the SUT, they will also send the code verifier
+(in plain text form), and Conjur will hash it and compare it to the stored
+code challenge.
+
+Initially we will support the SHA-256 algorithm for the code challenge. In the
+future we may want to support additional algorithms. We should require the
+client to specify the algorithm in the request headers, despite only supporting
+one algorithm for now, to allow for future expansion without breaking API
+compatibility.
+
+The code challenge and code verifier flow are modeled after OIDC PKCE.
+
+##### Example Database Table
+
+| role_id | token_hash | code_challenge | code_challenge_algorithm | expires_at |
+|---------|-----------|----------------|--------------------------|------------|
+| account:user:john.doe | lkduj...  | 3iod... | sha256 | 2023-09-28 12:08:23 |
+| account:user:demo.user | 2k1g2...  | kld80... | sha256 | 2023-09-28 12:09:45 |
+
+For this example, we would need a migration to add the `authn_sut_tokens` table.
+That would look something like this:
+
+```ruby
+Sequel.migration do
+  change do
+    create_table :authn_sut_tokens do
+      primary_key :token_hash, type: String
+      foreign_key :role_id, :roles, type: String, null: false, on_delete: :cascade
+      String :code_challenge, null: false
+      String :code_challenge_algorithm, null: false
+      DateTime :expires_at, null: false
+      index [:token_hash, :expires_at], unique: true
+    end
+  end
+end
+```
+
+#### Authenticator
+
+The SUT authenticator will be implemented as a new authenticator plugin. The
+implementation will follow the existing authenticator plugin pattern, as
+documented in [AUTHENTICATORS.md](../AUTHENTICATORS.md)
+
+## Security
+
+Due to the fact that this feature will be used to authenticate to Conjur, it is
+critical that the implementation is secure. Threat modeling should be performed
+before implementation begins to identify potential security issues and ensure
+that they are addressed. There should be an additional security review of the
+implementation before it is merged.
+
+Much of the security of this feature will be provided by the underlying
+authenticator architecture, so we only need to be concerned with the security of
+the SUT and the implementation of the SUT authenticator plugin.
+
+## Testing
```

Comment:
@szh nice list of test scenarios.  I had a couple comments/questions..

1) what if ldap is enabled for Conjur-UI login by default or authn-oidc?
2) check that tokens are not logged (even though they are short lived)
3) how about authn-temp as credential name?
4) will there be a status endpoint for authn_sut?
5) could you define more than one authn_sut authenticator?

Reply:
I'll clarify these. I think authn-temp is too vague - it makes it sound like the authenticator itself is temporary.

Reply:
I'm actually not sure about authn-ldap - I'll need some clarification from someone more familiar with UI project.